### PR TITLE
Racy `cancel` from transport does not always cancel the message body

### PR DESCRIPTION
**Pull request by idelpivnitskiy**
Motivation:

When `cancel` from the server's transport races with HTTP message completion, we can receive `cancel` after `Single<HttpResponseMetaData>` completes but before we subscribed to the `messageBody` publisher. In this case, we won't subscribe to the `messageBody` because we use `concat` operator. This does not allow `BeforeFinallyHttpOperator` to observe a terminal event. Race is possible when offloading is enabled. `cancel` comes from an IO thread, HTTP meta-data completes on offloaded thread.

Modifications:

- For the server-side control flow, use the recently added `concatPropagateCancel` operator to guarantee propagation of the cancel for the entire flat HTTP stream;

Result:

`HttpLifecycleObserverTest.testClientCancelsRequestBeforeResponse()` does not race anymore.

### CODE REVIEW

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HeaderUtils.java [hunk 1]
```diff
@@ -219,8 +224,55 @@ static Publisher<Object> flatEmptyMessage(final HttpProtocolVersion protocolVers
...
+                        toSource(messageBody).subscribe(CancelImmediatelySubscriber.INSTANCE);
+                    }
+                }).concat(messageBody.ignoreElements());
+    }
+
+    static Publisher<Object> flatMessage(final HttpMetaData metadata, final Publisher<Object> messageBody,
+                                         final boolean deferSubscribe) {
+        // Because `concat` won't subscribe to the messageBody in case of cancellation or an error, we use
+        // afterFinally to guarantee messageBody sees cancel too. Otherwise, observers won't complete exchange.
+        return Single.<Object>succeeded(metadata).afterFinally(new SingleTerminalSignalConsumer<Object>() {
```
Comment (Scottmitch) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HeaderUtils.java:256]: `Single.succeeded` ignores cancellation, do we need the additional complexity in this case?

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HeaderUtils.java [hunk 2]
```diff
@@ -219,8 +224,55 @@ static Publisher<Object> flatEmptyMessage(final HttpProtocolVersion protocolVers
                         from(metadata, EmptyHttpHeaders.INSTANCE);
         return messageBody == empty() ? flatMessage :
                 // Subscribe to the messageBody publisher to trigger any applied transformations, but ignore its
-                // content because the PayloadInfo indicated it's effectively empty and does not contain trailers
-                flatMessage.concat(messageBody.ignoreElements());
+                // content because the PayloadInfo indicated it's effectively empty and does not contain trailers.
+                // Because `concat` won't subscribe to the messageBody in case of cancellation or an error, we use
+                // afterFinally to guarantee messageBody sees cancel too. Otherwise, observers won't complete exchange.
+                flatMessage.afterFinally(new TerminalSignalConsumer() {
```
[the remark this answers is not in the document]
Comment (Scottmitch) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HeaderUtils.java:230]: Is this what we are seeing the test failure?
```
RetryRequestWithNonRepeatablePayloadTest.protocol=HTTP_1, offloading=false

java.lang.AssertionError: Unexpected exception(s) encountered
	at io.servicetalk.concurrent.api.TestPublisher.checkSubscriberAndExceptions(TestPublisher.java:178)
	at io.servicetalk.concurrent.api.TestPublisher.onNext(TestPublisher.java:140)
	at io.servicetalk.http.netty.RetryRequestWithNonRepeatablePayloadTest.test(RetryRequestWithNonRepeatablePayloadTest.java:138)

...

Caused by: java.lang.IllegalStateException: Duplicate subscriber: io.servicetalk.concurrent.api.AutoOnSubscribePublisherSubscriberFunction$1@44388d88
	at io.servicetalk.concurrent.api.NonResubscribeablePublisherSubscriberFunction.apply(NonResubscribeablePublisherSubscriberFunction.java:37)
	at io.servicetalk.concurrent.api.NonResubscribeablePublisherSubscriberFunction.apply(NonResubscribeablePublisherSubscriberFunction.java:29)
```

[the remark this answers is not in the document]
Comment (Scottmitch) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HeaderUtils.java:230]: We can use https://github.com/apple/servicetalk/pull/2372 in this case instead.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HeaderUtils.java (hunk 1 above)
Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HeaderUtils.java]: Doesn't matter if the first source ignores cancel or not, the problem is that `Single.concat(Publisher)` won't subscribe to the publisher if it was canceled before the result of a `Single` is delivered.

After digging more, I understood that we don't need this complexity on the client-side at all. Request payload does not affect `BeforeFinallyHttpOperator` on the client and the `AbstractLifecycleObserverHttpFilter` waits for the request payload body only if transport subscribed to it before the response was cancelled. So, we are good. On the server-side, we still need to propagate `cancel`.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HeaderUtils.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HeaderUtils.java]: Good catch!
For `Publisher.concat(Completable)` we can use your new overload, thanks!
For `Single.concat(Publisher, false)`, neither `beforeFinnaly` (we can miss cancel) nor `afterFinally` (duplicate subscribe) help. I came up with a local `AtomicBoolean` to make sure we subscribe only once.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HeaderUtils.java (hunk 1 above)
Comment (Scottmitch) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HeaderUtils.java]: ack. I think we should be able to use this: https://github.com/apple/servicetalk/pull/2381

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/NettyHttpServer.java [hunk 3]
```diff
@@ -453,11 +456,27 @@ private static Publisher<Object> handleResponse(final HttpProtocolVersion protoc
...
-                if (emptyMessageBody(response, response.messageBody())) {
-                    flatResponse = flatEmptyMessage(protocolVersion, response, response.messageBody());
+                final Publisher<Object> messageBody = response.messageBody();
+                if (emptyMessageBody(response, messageBody)) {
+                    flatResponse = flatEmptyMessage(protocolVersion, response, messageBody, true);
                 } else {
-                    // Not necessary to defer subscribe to the messageBody because server does not retry responses
-                    flatResponse = Single.<Object>succeeded(response).concat(response.messageBody());
+                    // No need to wrap the state with `defer` because this method is executed inside `flatMapPublisher`.
+                    final AtomicBoolean messageBodySubscribed = new AtomicBoolean(false);
```
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/NettyHttpServer.java:464]: Comments above describe why `defer` is not required in both cases

[the remark this answers is not in the document]
Comment (Scottmitch) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/NettyHttpServer.java:464]: ah conditional based upon usage. we should be able to skip the caveats with the new `concatPropagateCancel` operators.

**Comment by idelpivnitskiy** [comment]
@Scottmitch rebased after your changes, ready for review

**Review by Scottmitch — APPROVED** [review]
Is the test failure related at all?

https://github.com/apple/servicetalk/runs/8686381887
```
HttpTransportObserverTest.serverHandlerError(HttpProtocol) [1] protocol=HTTP_1

Wanted but not invoked:
serverWriteObserver.itemFlushed();
-> at io.servicetalk.http.netty.HttpTransportObserverTest.testRequestResponse(HttpTransportObserverTest.java:243)

However, there were exactly 4 interactions with this mock:
serverWriteObserver.requestedToWrite(64L);
-> at io.servicetalk.transport.api.CatchAllTransportObserver$CatchAllWriteObserver.lambda$requestedToWrite$0(CatchAllTransportObserver.java:255)

serverWriteObserver.itemReceived(
    HTTP/1.1 500 Internal Server Error
);
-> at io.servicetalk.transport.api.CatchAllTransportObserver$CatchAllWriteObserver.lambda$itemReceived$1(CatchAllTransportObserver.java:260)

serverWriteObserver.onFlushRequest();
-> at io.servicetalk.transport.api.CatchAllTransportObserver.safeReport(CatchAllTransportObserver.java:306)

serverWriteObserver.itemWritten(
    HTTP/1.1 500 Internal Server Error
);
-> at io.servicetalk.transport.api.CatchAllTransportObserver$CatchAllWriteObserver.lambda$itemWritten$2(CatchAllTransportObserver.java:270)


	at app//io.servicetalk.http.netty.HttpTransportObserverTest.testRequestResponse(HttpTransportObserverTest.java:243)
	at app//io.servicetalk.http.netty.HttpTransportObserverTest.serverHandlerError(HttpTransportObserverTest.java:213)
	
```

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/AbstractStreamingHttpConnection.java [hunk 4]
```diff
@@ -166,12 +166,18 @@ public Single<StreamingHttpResponse> request(final StreamingHttpRequest request)
...
-                if (emptyMessageBody(request, request.messageBody())) {
-                    flatRequest = flatEmptyMessage(connectionContext().protocol(), request, request.messageBody());
+                final Publisher<Object> messageBody = request.messageBody();
+                // Do not propagate cancel to the messageBody if cancel arrives before meta-data completes. Client-side
+                // state machine does not depend on termination of the messageBody until after transport subscribes to
+                // it. It's preferable to avoid subscribe to the messageBody in case of cancellation to allow requests
+                // with non-replayable messageBody to retry.
+                if (emptyMessageBody(request, messageBody)) {
+                    flatRequest = flatEmptyMessage(connectionContext().protocol(), request, messageBody,
+                            /* propagateCancel */ false);
```
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/AbstractStreamingHttpConnection.java:176]: It doesn't hurt to have it either. I frequently read code on GitHub without IDE and share a direct link to interesting places.

**Comment by idelpivnitskiy** [comment]
The test failyre is unrelated, it's known flaky test #2040

**Pull request merged**